### PR TITLE
feat: generalize some functions in sp-trie

### PR DESCRIPTION
**Pull request by yjhmelody**
- Add `to_memory_db`/`to_compact_proof` to reduce clone for `StorageProof` When need to reuse it.
- Generalize some methods.
----
polkadot address: 15ouFh2SHpGbHtDPsJ6cXQfes9Cx1gEFnJJsJVqPGzBSTudr

### CODE REVIEW

File: client/service/src/client/client.rs [hunk 1]
```diff
@@ -1190,7 +1190,8 @@ where
 			state, size_limit, start_key,
 		)?;
 		// This is read proof only, we can use either LayoutV0 or LayoutV1.
-		let proof = sp_trie::encode_compact::<sp_trie::LayoutV0<HashFor<Block>>>(proof, root)
+		let proof = proof
+			.into_compact_proof::<sp_trie::LayoutV0<HashFor<Block>>>(root)
```
Comment (cheme) [inline client/service/src/client/client.rs:1194]:
```suggestion
			.into_compact_proof::<HashFor<Block>>(root)
```

File: client/service/src/client/client.rs [hunk 2]
```diff
@@ -1190,7 +1190,8 @@ where
 			state, size_limit, start_key,
 		)?;
 		// This is read proof only, we can use either LayoutV0 or LayoutV1.
```
Comment (cheme) [inline client/service/src/client/client.rs:1192]:
```suggestion
```

File: primitives/trie/src/storage_proof.rs [hunk 3]
```diff
@@ -54,10 +58,16 @@ impl StorageProof {
 		self.trie_nodes.is_empty()
 	}
 
+	/// Convert into an iterator over encoded trie nodes in lexicographical order constructed
+	/// from the proof.
+	pub fn iter_nodes(self) -> IntoIter<Vec<u8>> {
```
Comment (cheme) [inline primitives/trie/src/storage_proof.rs:63]: Could be rename `into_iter_nodes`

File: primitives/trie/src/storage_proof.rs [hunk 4]
```diff
@@ -54,10 +58,16 @@ impl StorageProof {
...
+		self.trie_nodes.into_iter()
+	}
+
 	/// Create an iterator over encoded trie nodes in lexicographical order constructed
 	/// from the proof.
-	pub fn iter_nodes(self) -> StorageProofNodeIterator {
-		StorageProofNodeIterator::new(self)
+	pub fn iter(&self) -> Iter<'_, Vec<u8>> {
```
Comment (cheme) [inline primitives/trie/src/storage_proof.rs:69]: Could be rename `iter_nodes`

File: primitives/trie/src/trie_codec.rs [hunk 5]
```diff
@@ -149,15 +149,15 @@ where
...
 	root: TrieHash<L>,
 ) -> Result<CompactProof, Error<TrieHash<L>, CError<L>>>
 where
 	L: TrieConfiguration,
+	DB: HashDBT<L::Hash, trie_db::DBValue> + hash_db::HashDBRef<L::Hash, trie_db::DBValue>,
 {
 	let mut child_tries = Vec::new();
-	let partial_db = proof.into_memory_db();
 	let mut compact_proof = {
 		let trie = crate::TrieDBBuilder::<L>::new(&partial_db, &root).build();
```
Comment (cheme) [inline primitives/trie/src/trie_codec.rs:162]:
```suggestion
		let trie = crate::TrieDBBuilder::<L>::new(partial_db, root).build();
```

File: primitives/trie/src/trie_codec.rs [hunk 6]
```diff
@@ -149,15 +149,15 @@ where
 /// Then parse all child trie root and compress main trie content first
 /// then all child trie contents.
 /// Child trie are ordered by the order of their roots in the top trie.
-pub fn encode_compact<L>(
-	proof: StorageProof,
+pub fn encode_compact<L, DB>(
+	partial_db: DB,
 	root: TrieHash<L>,
```
Comment (cheme) [inline primitives/trie/src/trie_codec.rs:154]:
```suggestion
	partial_db: &DB,
	root: &TrieHash<L>,
```

**Review by cheme — COMMENTED** [review]
Thanks for the PR.
I added a suggestion to switch  the naming of the iterator functions (PR already change api so sounds fine to me).

File: primitives/trie/src/storage_proof.rs [hunk 7]
```diff
@@ -114,6 +139,17 @@ impl<H: Hasher> From<StorageProof> for crate::MemoryDB<H> {
 	}
 }
 
+impl<H: Hasher> From<&StorageProof> for crate::MemoryDB<H> {
+	fn from(proof: &StorageProof) -> Self {
+		let mut db = crate::MemoryDB::default();
```
Comment (bkchr) [inline primitives/trie/src/storage_proof.rs:144]: Please make the implementation above use this one `(&self).into()` should do it.

File: primitives/trie/src/storage_proof.rs (hunk 3 above)
Comment (bkchr) [reply primitives/trie/src/storage_proof.rs:63]: Should be!

File: primitives/trie/src/storage_proof.rs (hunk 4 above)
Comment (bkchr) [reply primitives/trie/src/storage_proof.rs:69]: Yeah, this rename doesn't make any sense.

File: primitives/trie/src/storage_proof.rs (hunk 3 above)
Comment (bkchr) [inline primitives/trie/src/storage_proof.rs:63]: You can also just use `impl IntoIter` as return value.

File: primitives/trie/src/storage_proof.rs (hunk 4 above)
Comment (bkchr) [reply primitives/trie/src/storage_proof.rs:69]: And here return `impl Iter`

**Review by bkchr — COMMENTED** [review]
Thank you. 

Left some more comments.

File: primitives/trie/src/trie_codec.rs (hunk 6 above)
Comment (yjhmelody) [reply primitives/trie/src/trie_codec.rs:154]: If use`DB` ref here, Must use `dyn` here, like `TrieDBBuilder`

File: primitives/trie/src/storage_proof.rs (hunk 4 above)
Comment (yjhmelody) [reply primitives/trie/src/storage_proof.rs:69]: > And here return `impl Iter`

Yeah, But I think user maybe wanna use more trait methods here.

Comment (bkchr) [reply primitives/trie/src/storage_proof.rs]: Sounds good.

Comment (bkchr) [reply primitives/trie/src/storage_proof.rs]: We didn't supported this before. So, not sure we need this now.

**Comment by yjhmelody** [comment]
Hi, any else need to be improved? @bkchr

**Review by bkchr — APPROVED** [review]

**Review by cheme — APPROVED** [review]
Sorry for the delay. Merge with master seems needed before this can get merge.

**Comment by bkchr** [comment]
bot rebase

**Comment by bkchr** [comment]
bot merge

**Pull request merged**